### ISSUE TITLE
[FEA] support more than one batching dims in batched_gemm

### ISSUE THREAD

**Issue by minminsun** [NVIDIA/cutlass#454]
Batched_gemm is the key operation in attention computation.
In multi-head attention, there are 2 batching dims, a.k.a. batch_size dim and num_head dim. And the tensor layout before multi-head attention computation is [**batch_size**, seq_length, **num_head**, hidden_size].
Currently batched_gemm in either cublas or cutlass supports only one batching dim (please correct me if I'm wrong). Multiple batching dims have to get combined to call batched_gemm. Only contiguous batching dims can be combined so a transpose operation is needed ([**batch_size**, seq_length, **num_head**, hidden_size] -> [**batch_size**, **num_head**, seq_length, hidden_size] ).
If batched_gemm in cutlass can be extended to support more than one batching dims, the transpose can be eliminated. There are totally 4 transpose operations (3 for q/k/v inputs and 1 for output) can be eliminated in a multi-head attention layer.
An alternative way is to use grouped_gemm. But it requires a memcpy from host to device for the pointer array. So it would be very helpful is cutlass supports batched_gemm with multiple noncontiguous batching dims.
Thanks!

**Comment by hwu36** [comment]
Ggroup gemm is the way to go.  The pointer array should not be a big problem.
1. The array size is small.  
2. Groups do not need to be contiguous.  You can set the point array much earlier than the group gemm kernel.

**Comment by minminsun** [comment]
> Ggroup gemm is the way to go. The pointer array should not be a big problem.
> 
> 1. The array size is small.
> 2. Groups do not need to be contiguous.  You can set the point array much earlier than the group gemm kernel.

Thanks for your reply. grouped_gemm is one option indeed, but I have some concern about it.
Although the pointer array size is very small, it may still take at least a kernel launch time (~5 us), right? For bert model with small batch size, the transpose kernel takes only ~5 us too. Some the time saving of transpose removal may get discounted in such cases.
I don't quite understand "You can set the point array much earlier than the group gemm kernel". The pointers are available only after the input/output tensors have been allocated. In both pytorch and tensorflow, the tensor allocation doesn't happen too earlier than op kernel launch.

**Comment by hwu36** [comment]
Try it :).  It should work.  See https://blogs.nvidia.com/blog/2022/03/22/microsoft-translator-triton-inference/

**Comment by mnicely** [comment]
@minminsun were you able to use grouped gemm?
